conn.go

```diff
@@ -900,7 +900,12 @@ func (c *Conn) queueRequest(opcode int32, req interface{}, res interface{}, recv
 	switch opcode {
 	case opClose:
 		// always attempt to send close ops.
-		c.sendChan <- rq
+		select {
+		case c.sendChan <- rq:
+		case <-time.After(time.Second):
+			c.logger.Printf("gave up trying to send opClose to server")
+			rq.recvChan <- response{-1, ErrConnectionClosed}
+		}
 	default:
 		// otherwise avoid deadlocks for dumb clients who aren't aware that
 		// the ZK connection is closed yet.
```

Comment:
overall i think this comment sums up the notion that this code fix for a deadlock can be avoided with proper client logic.

Can you provide a reproduction Unit test to demonstrate the case you are fixing.

Reply:
I ask since we run massive deployments with thousands of connections and have no issue in this area. i think your analysis is correct but i would rather lean in on avoiding more implicit timeouts. if we could indeed thread in a context in these private methods it may be better.

Reply:
Not sure what a fix to a client's logic whose goroutines are blocked inside this library would be. Mind to elaborate your concern?

As for reproducibility in form of a unit test, I wish I could promise you that. But I can't. The description of issue #13  which this PR is an attempt to solve should give a feel of the conditions needed for the deadlock. I don't see a way to express them as a test.

Reply:
`queueRequest` is quite testable since you can create and fully control `c.sendChan`  on a `Conn`. 

I would create a testing `Conn` that doesn't read this channel and you can get a deadlock case in your unit test. To assert this condition you would wait a set amount of time (say a second) and assert it never reached the return.

In the test it's a bit complex since you need to start a groutine that is suppose to deadlock and assert that it does in fact deadlock.

If we then add a context you can cancel the context as a test and assert we immediately return.

Reply:
@jeffbean Thanks for taking your time reviewing this proposal. I totally agree that re-wiring those in favour of context-based cancellation would be preferable. Though, that comes with a totally different resource tag, so I'm leaving that open for others to take on.

Reply:
> `queueRequest` is quite testable since you can create and fully control `c.sendChan` on a `Conn`.
> 
> I would create a testing `Conn` that doesn't read this channel and you can get a deadlock case in your unit test. To assert this condition you would wait a set amount of time (say a second) and assert it never reached the return.
> 
> In the test it's a bit complex since you need to start a groutine that is suppose to deadlock and assert that it does in fact deadlock.

That does make sense, thanks – I'll give it a try.

Reply:
> @jeffbean Thanks for taking your time reviewing this proposal. I totally agree that re-wiring those in favour of context-based cancellation would be preferable. Though, that comes with a totally different resource tag, so I'm leaving that open for others to take on.

awesome - i agree and we can do contextual changes later. carry on in this direction. :D

Reply:
Added a test - thanks for the advice!